### PR TITLE
CIRC-1371 Post to /circulation/requests without request date returns error but changes item status

### PR DESCRIPTION
**Pull request by roman-barannyk**
## Purpose
Fix the bug regarding creating a request without a request date the service returns an error and does NOT create the request but it does change the item status to Paged (for a previously 'available' item)

Resolves: [CIRC-1371](https://issues.folio.org/browse/CIRC-1371)

## Approach
- Create requestDate validation;
- Create test;

### CODE REVIEW

File: src/test/java/api/requests/RequestsAPICreationTests.java [hunk 1]
```diff
@@ -2835,6 +2836,31 @@ void statusOfTlrRequestShouldBeChangedIfAssociatedItemCheckedIn() {
     assertThat(requestAssociatedWithFirstItem.getString("status"), is(OPEN_AWAITING_PICKUP));
   }
 
+  @Test
+  void pageRequestShouldNotChangeItemStatusIfFailsWithoutRequestDate() {
```
Comment (viktor-draban) [inline src/test/java/api/requests/RequestsAPICreationTests.java:2840]:
```suggestion
  void pageRequestShouldNotChangeItemStatusIfRequestDateNone() {
```

File: src/main/java/org/folio/circulation/resources/RequestFromRepresentationService.java [hunk 2]
```diff
@@ -341,6 +345,21 @@ private Request addUserToLoan(Request request, User user) {
...
+  private Result<Request> refuseWhenNoRequestDate(Result<Request> request) {
+    return request.next(this::validateIfRequestDateIsPresent)
+      .mapFailure(err -> errorHandler.handleValidationError(err, INVALID_ITEM_ID, request));
+  }
+
+  private Result<Request> validateIfRequestDateIsPresent(Request context) {
+    var requestDate = getDateTimeProperty(context.getRequestRepresentation(), REQUEST_DATE);
+
+    if (requestDate == null) {
+      return failedValidation("Cannot create a request with no requestDate", REQUEST_DATE, null);
```
Comment (viktor-draban) [inline src/main/java/org/folio/circulation/resources/RequestFromRepresentationService.java:357]:
```suggestion
      return failedValidation("Cannot create a request without requestDate", REQUEST_DATE, null);
```

File: src/test/java/api/requests/RequestsAPICreationTests.java [hunk 3]
```diff
@@ -2835,6 +2836,31 @@ void statusOfTlrRequestShouldBeChangedIfAssociatedItemCheckedIn() {
...
+    UUID pickupServicePointId = servicePointsFixture.cd1().getId();
+    var item = itemsFixture.basedUponSmallAngryPlanet();
+    var requester = usersFixture.steve();
+
+    Response response = requestsClient.attemptCreate(new RequestBuilder()
+      .page()
+      .fulfilToHoldShelf()
+      .withRequesterId(requester.getId())
```
Comment (viktor-draban) [inline src/test/java/api/requests/RequestsAPICreationTests.java:2848]:
```suggestion
      .withRequesterId(usersFixture.steve()getId())
```
I'm not sure we need variable requester

File: src/test/java/api/requests/RequestsAPICreationTests.java [hunk 4]
```diff
@@ -2835,6 +2836,31 @@ void statusOfTlrRequestShouldBeChangedIfAssociatedItemCheckedIn() {
...
+      .withItemId(item.getId())
+      .withPickupServicePointId(pickupServicePointId)
```
Comment (viktor-draban) [inline src/test/java/api/requests/RequestsAPICreationTests.java:2850]:
```suggestion
      .withPickupServicePointId(servicePointsFixture.cd1().getId())
```
the same

File: src/test/java/api/requests/RequestsAPICreationTests.java [hunk 5]
```diff
@@ -2835,6 +2836,31 @@ void statusOfTlrRequestShouldBeChangedIfAssociatedItemCheckedIn() {
...
+      .withInstanceId(item.getInstanceId())
+      .itemRequestLevel()
+      .withHoldingsRecordId(item.getHoldingsRecordId()).withRequestDate(null));
```
Comment (viktor-draban) [inline src/test/java/api/requests/RequestsAPICreationTests.java:2853]:
```suggestion
      .withHoldingsRecordId(item.getHoldingsRecordId())
      .withRequestDate(null));
```
let's stick on one style

File: src/test/java/api/requests/RequestsAPICreationTests.java [hunk 6]
```diff
@@ -2835,6 +2836,31 @@ void statusOfTlrRequestShouldBeChangedIfAssociatedItemCheckedIn() {
...
+
+    assertThat(response, hasStatus(HTTP_UNPROCESSABLE_ENTITY));
+    assertThat(response.getJson(), hasErrorWith(allOf(
+      hasMessage("Cannot create a request with no requestDate"),
```
Comment (viktor-draban) [inline src/test/java/api/requests/RequestsAPICreationTests.java:2857]:
```suggestion
      hasMessage("Cannot create a request without requestDate"),
```

File: src/main/java/org/folio/circulation/resources/RequestFromRepresentationService.java (hunk 2 above)
Comment (roman-barannyk) [reply src/main/java/org/folio/circulation/resources/RequestFromRepresentationService.java:357]: I used the same style as it had been already been in the same class. Please see the other error messages above: 
 "Cannot create a request with no instance ID", "Cannot create an item level request with no item ID" etc.

File: src/main/java/org/folio/circulation/resources/RequestFromRepresentationService.java [hunk 7]
```diff
@@ -341,6 +345,21 @@ private Request addUserToLoan(Request request, User user) {
...
+    } else {
+      return of(() -> context);
+    }
+  }
+
```
[the remark this answers is not in the document]
Comment (roman-barannyk) [reply src/main/java/org/folio/circulation/resources/RequestFromRepresentationService.java:362]: I put this methods right after the similar which is called from the the chain above. In terms of refactoring with moving validation to another class -  I don't think we need to do it in the scope of this bug

File: src/test/java/api/requests/RequestsAPICreationTests.java (hunk 3 above)
Comment (roman-barannyk) [reply src/test/java/api/requests/RequestsAPICreationTests.java]: done

File: src/test/java/api/requests/RequestsAPICreationTests.java (hunk 4 above)
Comment (roman-barannyk) [reply src/test/java/api/requests/RequestsAPICreationTests.java]: done

File: src/test/java/api/requests/RequestsAPICreationTests.java (hunk 5 above)
Comment (roman-barannyk) [reply src/test/java/api/requests/RequestsAPICreationTests.java]: sure, thanks

File: src/test/java/api/requests/RequestsAPICreationTests.java (hunk 1 above)
Comment (roman-barannyk) [reply src/test/java/api/requests/RequestsAPICreationTests.java:2840]: I would leave it as it is

File: src/main/java/org/folio/circulation/resources/RequestFromRepresentationService.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (viktor-draban) [reply src/main/java/org/folio/circulation/resources/RequestFromRepresentationService.java:362]: Method above is refuseToCreateTlrLinkedToAnItem
which calls -> refuseToCreateTlrLinkedToAnItem and then -> validateAbsenceOfItemLinkInTlr

But your methods (refuseWhenNoRequestDate validateIfRequestDateIsPresent) between them, am I right ?

File: src/main/java/org/folio/circulation/resources/RequestFromRepresentationService.java (hunk 2 above)
Comment (viktor-draban) [reply src/main/java/org/folio/circulation/resources/RequestFromRepresentationService.java:357]: got it

File: src/test/java/api/requests/RequestsAPICreationTests.java (hunk 1 above)
Comment (viktor-draban) [reply src/test/java/api/requests/RequestsAPICreationTests.java:2840]: ok

**Review by viktor-draban — APPROVED** [review]

**Review by bisecomsergiy — APPROVED** [review]

**Review by alexanderkurash — APPROVED** [review]

**Pull request merged**